File: code_snippets/java/src/main/java/usecases/microservices/SubscriptionService.java
package usecases.microservices;

import static usecases.microservices.Utils.createRecurringPayment;
import static usecases.microservices.Utils.createSubscription;

import dev.restate.sdk.Context;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.http.vertx.RestateHttpServer;
import usecases.microservices.Utils.SubscriptionRequest;

// <start_here>
// <mark_1>
@Service
public class SubscriptionService {
  // </mark_1>

  // <mark_1>
  @Handler
  public void add(Context ctx, SubscriptionRequest req) {
    // </mark_1>
    var paymentId = ctx.random().nextUUID().toString();

    // <mark_2>
    var payRef =
        ctx.run(
            String.class,
            // break
            () -> createRecurringPayment(req.creditCard(), paymentId));
    // </mark_2>

    for (String subscription : req.subscriptions()) {
      // <mark_2>
      ctx.run(() -> createSubscription(req.userId(), subscription, payRef));
      // </mark_2>
    }
  }

  public static void main(String[] args) {
    RestateHttpServer.listen(
        // break
        Endpoint.bind(new SubscriptionService()));
  }
}
// <end_here>

File: code_snippets/python/src/concepts/services/subscription_service.py
import uuid

import restate
from pydantic import BaseModel
from restate import Service, Context


class SubscriptionRequest(BaseModel):
    user_id: str
    credit_card: str
    subscriptions: list[str]


# <start_here>
# <mark_2>
subscription_service = Service("SubscriptionService")
# </mark_2>


# <mark_2>
@subscription_service.handler()
async def add(ctx: Context, req: SubscriptionRequest):
    # <mark_3>
    # </mark_2>
    # <mark_1>
    payment_id = await ctx.run("payment id", lambda: str(uuid.uuid4()))
    # </mark_1>

    # <mark_1>
    pay_ref = await ctx.run(
        "recurring payment",
        lambda: create_recurring_payment(req.credit_card, payment_id),
    )
    # </mark_1>

    for subscription in req.subscriptions:
        # <mark_1>
        await ctx.run(
            "subscription",
            lambda: create_subscription(req.user_id, subscription, pay_ref),
        )
        # </mark_1>
    # </mark_3>


# <mark_4>
app = restate.app([subscription_service])
# </mark_4>
# <end_here>


def create_recurring_payment(credit_card, payment_id):
    return None


def create_subscription(user_id, subscription, pay_ref):
    return None
